src/D3MHub.sol

```diff
@@ -395,126 +390,44 @@ contract D3MHub {
         @param ilk bytes32 of the D3M ilk name
     */
     function exec(bytes32 ilk) external lock {
+        // IMPORTANT: this function assumes Vat rate of D3M ilks will always be == 1 * RAY (no fees).
+        // That's why this module converts normalized debt (art) to Vat DAI generated with a simple RAY multiplication or division
+
         (uint256 Art, uint256 rate, uint256 spot, uint256 line,) = vat.ilks(ilk);
         require(rate == RAY, "D3MHub/rate-not-one");
         require(spot == RAY, "D3MHub/spot-not-one");
 
         ID3MPool _pool = ilks[ilk].pool;
 
-        _pool.preDebtChange("exec");
-        uint256 currentAssets = _pool.assetBalance();
+        _pool.preDebtChange();
 
         if (vat.live() == 0) {
             // MCD caged
-            require(end.debt() == 0, "D3MHub/end-debt-already-set");
+            EndLike _end = end;
+            require(_end.debt() == 0, "D3MHub/end-debt-already-set");
             require(ilks[ilk].culled == 0, "D3MHub/module-has-to-be-unculled-first");
+            _end.skim(ilk, address(_pool));
```

Comment:
Have we thought through the cases where someone else calls `End.skim` for the pool permissionlessly? At a glance, it should be fine if `culled == 0` b/c `skim` is idempotent, but it could be a problem if `culled == 1`.

Reply:
Missed Tal response when replying. Basically what he says.

Reply:
Okay think we're good here.